flocker/control/_model.py

```diff
@@ -414,6 +425,26 @@ def key_match_invariant(pmap):
 _keys_match_dataset_id = _keys_match("dataset_id")
 
 
+def _turn_iterable_to_mapping_from_attribute(attribute, obj):
+    """
+    Given ``obj`` which is either something that is an instance of ``Mapping``
+    or an iterable, return something that implements ``Mapping`` from
+    ``getattr(o, attribute)`` to the value of the items in the iterable. This
+    funky signature is required to be a valid field factory function.
+
+    :param attribute: The name of the attribute that should be the key in the
+        resulting Mapping.
+    :param obj: Either something that is a Mapping (which is immediately
+        returned), or an iterable to be converted into a Mapping.
+
+    :returns: A Mapping of the items in ``obj`` from
+        ``getattr(o, attribute)`` to the items in obj (``o`` in ``obj``).
+    """
+    if isinstance(obj, Mapping):
+        return obj
+    return {getattr(a, attribute): a for a in obj}
```

Comment:
I guess it doesn't matter too much, but this gives a non-obvious error if you accidentally assign e.g. a list or set


```
In [81]: Foo().transform(['bar'], {"abc", 123})
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-81-90b6a18eed00> in <module>()
----> 1 Foo().transform(['bar'], {"abc", 123})

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_pclass.pyc in transform(self, *transformations)
    133         for other persistent types work as expected.
    134         """
--> 135         return transform(self, transformations)
    136 
    137     def __eq__(self, other):

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_transformations.pyc in transform(structure, transformations)
     41     r = structure
     42     for path, command in _chunks(transformations, 2):
---> 43         r = _do_to_path(r, path, command)
     44     return r
     45 

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_transformations.pyc in _do_to_path(structure, path, command)
     50 
     51     kvs = _get_keys_and_values(structure, path[0])
---> 52     return _update_structure(structure, kvs, path[1:], command)
     53 
     54 

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_transformations.pyc in _update_structure(structure, kvs, path, command)
     89             if result is not v:
     90                 e[k] = result
---> 91     return e.persistent()

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_pclass.pyc in persistent(self)
    229     def persistent(self):
    230         if self._pclass_evolver_data_is_dirty:
--> 231             return self._pclass_evolver_original.__class__(**self._pclass_evolver_data)
    232 
    233         return self._pclass_evolver_original

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_pclass.pyc in __new__(cls, **kwargs)
     50         for name, field in cls._pclass_fields.items():
     51             if name in kwargs:
---> 52                 value = field.factory(kwargs[name])
     53                 _check_and_set_attr(cls, field, name, value, result, invariant_errors)
     54                 del kwargs[name]

/home/richard/projects/HybridLogic/flocker/flocker/control/_model.py in <lambda>(x)
    136 
    137     if input_factory:
--> 138         factory = lambda x: mapping_factory(input_factory(x))
    139     else:
    140         factory = mapping_factory

<ipython-input-69-59b958d20f7e> in <lambda>(x)
      1 class Foo(PClass):
----> 2     bar = pmap_field(key_type=bytes, value_type=int, factory=lambda x: _turn_iterable_to_mapping_from_attribute('name', x))
      3 

/home/richard/projects/HybridLogic/flocker/flocker/control/_model.py in _turn_iterable_to_mapping_from_attribute(attribute, obj)
    443     if isinstance(obj, Mapping):
    444         return obj
--> 445     return {getattr(a, attribute): a for a in obj}
    446 
    447 

/home/richard/projects/HybridLogic/flocker/flocker/control/_model.py in <dictcomp>((a,))
    443     if isinstance(obj, Mapping):
    444         return obj
--> 445     return {getattr(a, attribute): a for a in obj}
    446 
    447 

AttributeError: 'int' object has no attribute 'name'

```


Maybe the function could  do a little extra validation of the supplied objects and raise a meaningful error message.